### PR TITLE
Symbolic psd constraint

### PR DESCRIPTION
**Pull request by hongkai-dai**
Enforce a matrix of linear expression being positive semidefinite.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/5018)

### CODE REVIEW

**Comment by hongkai-dai** [comment]
+@sammy-tri for feature review please.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-:-KbvYiOzl3gzLLag0NML:bjquaum)*

**Comment by hongkai-dai** [comment]
@drake-jenkins-bot linux-xenial-gcc-ninja-experimental-open-source-debug please

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-:-KbvdNGdqEYcwd6Pv5kL:b-nikqms)*

**Comment by sammy-tri** [comment]
+(status: curate commits before merging) 

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1790 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-KbzL5iF0PtziIJOX9hA:-KbzL5iF0PtziIJOX9hB:b-3yyi02) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/mathematical_program.h#L1790)):*
> ```objc
>       std::is_same<typename Derived::Scalar, symbolic::Expression>::value,
>       Binding<PositiveSemidefiniteConstraint>>::type
>   AddPositiveSemidefiniteConstraint(const Eigen::MatrixBase<Derived>& e) {
> ```

I'm not clear why we're using an enable_if on a templated method instead of an overload like "AddPositiveSemidefiniteConstraint(const Eigen::MatrixBase<symbolic::Expression>&);"

---

*[drake/solvers/test/mathematical_program_test.cc, line 1032 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-KbzLRL8plgGjJ5zVo3I:-KbzLRL9f8cJHngJuOaX:b-rz3a3z) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1032)):*
> ```cpp
> namespace {
> template <typename Derived>
> typename std::enable_if<
> ```

See  comment from header about enable_if  + Derived vs. non-templated overload

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-:-KbzHpVjHNa6fPIxD_uG:b-s794vx)*

**Comment by hongkai-dai** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1790 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-KbzL5iF0PtziIJOX9hA:-KbzeUMz_BxeDoGeuO2T:b-yxki77) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/mathematical_program.h#L1790)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I'm not clear why we're using an enable_if on a templated method instead of an overload like "AddPositiveSemidefiniteConstraint(const Eigen::MatrixBase<symbolic::Expression>&);"
</blockquote></details>

I wrote it as a templated version, to avoid some memory allocation. Inside this function, I create a new matrix `M` and `M_flat`. The size of these two matrices can be known at compile time if I use the templated version, but I cannot do that with `Eigen::Ref<MatrixXDecisionVariable>`.

This can be a pre-mature optimization. If you think it is unnecessary, I can change to a non-templated version.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018)*

**Comment by sammy-tri** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/mathematical_program.h, line 1790 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-KbzL5iF0PtziIJOX9hA:-KbzhH1c-jHAd8k1DLXC:bjhpx4q) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/mathematical_program.h#L1790)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I wrote it as a templated version, to avoid some memory allocation. Inside this function, I create a new matrix `M` and `M_flat`. The size of these two matrices can be known at compile time if I use the templated version, but I cannot do that with `Eigen::Ref<MatrixXDecisionVariable>`.

This can be a pre-mature optimization. If you think it is unnecessary, I can change to a non-templated version.
</blockquote></details>

Nope, leave it as it is.  I got it the first time then just looked at the signature when I came back to look again after reading the test.   Sorry for the confusion.  (though I'm still not sure the enable_if is required on the test?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-:-Kbzhl-VpcwgHppiAh68:bnfp4nl)*

**Comment by hongkai-dai** [comment]
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[drake/solvers/mathematical_program.h, line 1790 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-KbzL5iF0PtziIJOX9hA:-KbzqNW9-8IYQmViYhUh:byg7811) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/mathematical_program.h#L1790)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Nope, leave it as it is.  I got it the first time then just looked at the signature when I came back to look again after reading the test.   Sorry for the confusion.  (though I'm still not sure the enable_if is required on the test?)
</blockquote></details>

I put the `enable_if` there, since we have two calls of `AddPositiveSemidefiniteConstraint`, by adding either a matrix of variables, or a matrix of expressions. Namely we have two functions
```C++
AddPositiveSemidefiniteConstraint(const Eigen::Ref<const MatrixXDecisionVariable>& V); // V is a matrix of symbolic variables.
template <typename Derived>
std::enable_if<std::is_same<typename Derived::Scalar, symbolic::Variable>::value, Binding<PositiveSemidefiniteConstraint>>::type
AddPositiveSemidefiniteConstraint(const Eigen::MatrixBase<Derived>& V); // V is a matrix of symbolic expressions
```
The `enable_if` is used to differentiate the call on a matrix of symbolic variables from a matrix of symbolic expressions. Inside these two `AddPositiveSemidefiniteConstraint` functions, the implementations are different. The first one just add a `PositiveSemidefiniteConstraint` directly, while the second function creates a new matrix of symbolic variables, imposes the linear equality constraint that the symbolic expression equals to the new symbolic variable, and then add the constraint that the new symbolic matrix of variables is positive semidefinite.

In the test, that function is only meant to test the call on the `AddPositiveSemidefiniteConstraint` with a symbolic expression.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018)*

**Comment by hongkai-dai** [comment]
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-:-KbzsrBn8RZdr0z2VStF:b-l70wzf)*

**Comment by hongkai-dai** [comment]
+@david-german-tri for platform review

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-:-KbztCv40b7huNsy1he2:b-rhk18b)*

**Comment by david-german-tri** [comment]
+(status: curate commits before merging) still appears needed to me

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-:-KbzwRb4nDceY0waPjaH:b-igm3zl)*

**Comment by david-german-tri** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1204 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-TjpgoqC-XXqNOvI3:-Kc-Tjphi-E_KzHIhZMH:bmlywne) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/mathematical_program.h#L1204)):*
> ```objc
>    *           1, 1;
>    *   Eigen::Vector2d beq(1, 3);
>    *   prog.AddLinearEqualityConstraint(Aeq, beq,{x.segment<1>(2), x.segment<1>(5)});
> ```

BTW `>80`

---

*[drake/solvers/test/mathematical_program_test.cc, line 1050 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-f5VPEQfUqb2lI6iE:-Kc-f5VQzcrTApS6Sae1:b-yana5v) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1050)):*
> ```cpp
>   // Check if the added linear constraint is correct. M is the newly added
>   // variables representing the psd matrix.
>   const Eigen::Map<const MatrixX<Variable>> M(&binding.variables()(0), V.rows(),
> ```

Is there a need to assert on the dimensions before doing this?

---

*[drake/solvers/test/mathematical_program_test.cc, line 1052 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-fByrRsWH1cnzOBZJ:-Kc-fByrRsWH1cnzOBZK:b-394uhe) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1052)):*
> ```cpp
>   const Eigen::Map<const MatrixX<Variable>> M(&binding.variables()(0), V.rows(),
>                                               V.cols());
>   auto V_minus_M = math::ToSymmetricMatrixFromLowerTriangularColumns(
> ```

This needs some kind of additional comment; I don't understand why `ToSymmetricMatrix...`

---

*[drake/solvers/test/mathematical_program_test.cc, line 1053 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-bSK0tajvKlrZ7l45:-Kc-bSK0tajvKlrZ7l46:bwvsyn7) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1053)):*
> ```cpp
>                                               V.cols());
>   auto V_minus_M = math::ToSymmetricMatrixFromLowerTriangularColumns(
>       prog->linear_equality_constraints().back().constraint()->A() *
> ```

BTW consider a temporary const reference for the result of `back()`.

---

*[drake/solvers/test/mathematical_program_test.cc, line 1055 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-hqYH33-kqAmZomn5:-Kc-hqYH33-kqAmZomn6:b-d2qxzt) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1055)):*
> ```cpp
>       prog->linear_equality_constraints().back().constraint()->A() *
>           prog->linear_equality_constraints().back().variables() -
>       prog->linear_equality_constraints().back().constraint()->lower_bound());
> ```

BTW lower_bound() == upper_bound()`, right?  Have you considered an additional accessor called `bound()` or `value()` or something?

---

*[drake/solvers/test/mathematical_program_test.cc, line 1067 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-c24vLtYyd7a_jkRl:-Kc-c24w8HZN_FyjAI15:b-kshmgi) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1067)):*
> ```cpp
>   EXPECT_EQ(prog.positive_semidefinite_constraints().size(), 1);
>   EXPECT_EQ(psd_cnstr.get(),
>             prog.positive_semidefinite_constraints().back().constraint().get());
> ```

BTW a reference might be useful here too

---

*[drake/solvers/test/mathematical_program_test.cc, line 1076 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-hGWYx1Wb8PN_FM8O:-Kc-hGWYx1Wb8PN_FM8P:b-ezviwg) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1076)):*
> ```cpp
>       &prog, Matrix4d::Identity() * X);
> 
>   // Checks 2 * X + E is psd.
> ```

E?  Did you mean I?

---

*[drake/solvers/test/mathematical_program_test.cc, line 1080 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-h7D_wGPPODiYN5iD:-Kc-h7D_wGPPODiYN5iE:bcy7g5j) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1080)):*
> ```cpp
>       &prog, 2.0 * X + Matrix4d::Identity());
> 
>   // Checks Aᵀ * X + X * A is psd.
> ```

If I'm understanding correctly, this is a general result for triangular A (Cholesky decomposition), right?  If so, please say so in a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018)*

**Comment by hongkai-dai** [comment]
Review status: 1 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1204 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-TjpgoqC-XXqNOvI3:-Kc-zgfCAqnzOQgAa0oc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/mathematical_program.h#L1204)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW `>80`
</blockquote></details>

Done.

---

*[drake/solvers/test/mathematical_program_test.cc, line 1050 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-f5VPEQfUqb2lI6iE:-Kc-stSdLsBg_w1v4HVa:bvk5t40) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1050)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Is there a need to assert on the dimensions before doing this?
</blockquote></details>

It is checked in AddPositiveSemidefiniteConstraint, that `V.rows() == V.cols()`.

---

*[drake/solvers/test/mathematical_program_test.cc, line 1052 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-fByrRsWH1cnzOBZJ:-Kc0-ETnVZOcfZTK74CK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1052)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

This needs some kind of additional comment; I don't understand why `ToSymmetricMatrix...`
</blockquote></details>

Done.

---

*[drake/solvers/test/mathematical_program_test.cc, line 1053 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-bSK0tajvKlrZ7l45:-Kc0-UEHftklyB3jYek0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1053)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW consider a temporary const reference for the result of `back()`.
</blockquote></details>

Done.

---

*[drake/solvers/test/mathematical_program_test.cc, line 1055 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-hqYH33-kqAmZomn5:-Kc-r-Nf4HX898kZJHBH:b-7ak6ii) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1055)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW lower_bound() == upper_bound()`, right?  Have you considered an additional accessor called `bound()` or `value()` or something?
</blockquote></details>

Yes `lower_bound() == upper_bound()`. I will add a function `bound()` in `LinearEqualityConstraint()` in a separate PR.

---

*[drake/solvers/test/mathematical_program_test.cc, line 1067 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-c24vLtYyd7a_jkRl:-Kc0-m6gdqximrxFaMDZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1067)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW a reference might be useful here too
</blockquote></details>

Done.

---

*[drake/solvers/test/mathematical_program_test.cc, line 1076 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-hGWYx1Wb8PN_FM8O:-Kc-qPz86EkZf_uQ7oIj:b-ckc9s7) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1076)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

E?  Did you mean I?
</blockquote></details>

In dynamics team `I` is used to mean inertia, and @sherm1 uses `E` to mean identity.

---

*[drake/solvers/test/mathematical_program_test.cc, line 1080 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-Kc-h7D_wGPPODiYN5iD:-Kc-osmB5-prNLgHOnJf:b-iklfh7) ([raw file](https://github.com/robotlocomotion/drake/blob/82fdec7d248d2ec7244cb885b58d4e8f66466057/drake/solvers/test/mathematical_program_test.cc#L1080)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

If I'm understanding correctly, this is a general result for triangular A (Cholesky decomposition), right?  If so, please say so in a comment.
</blockquote></details>

The general result for triangular A requires stronger condition. The theory is that for a triangular `A`, if all its diagonal entries are non-negative, then there exists a positive semidefinite matrix `X`, that guarantees `A' * X + X * A` is positive semidefinite. For example, for matrix `A = -identity()`, there does not exist any positive semidefinite matrix `X` to make `A' * X + X * A` being positive semidefinite.

But I am not solving an optimization problem to check if such `X` exists or not. The test here is just to adding this positive semidefinite constraint into `MathematicalProgram`, and make sure `MathematiaclProgram` parses the constraint correctly, so that we know `AddPositiveSemidefiniteConstraint` can parse a matrix of symbolic expressions. There is a separate test in `convex_optimization_test` that actually solves this optimization program, and checks if `X` in the solution is a positive semidefinite matrix.

`mathematical_program_test` does not require any solvers (Gurobi, Mosek, etc). This test only makes sure that functions like adding variables/cost/constraint work correctly in `MathematicalProgram`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018)*

**Comment by david-german-tri** [comment]
:lgtm:

---

Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-:-Kc0AiHEkX4Na16mTCAv:bnfp4nl)*

**Comment by hongkai-dai** [comment]
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5018#-:-Kc0B9fGu8qY1sUuhVKN:b-l70wzf)*

File: drake/solvers/test/convex_optimization_test.cc [hunk 1]
```diff
@@ -1284,42 +1284,14 @@ GTEST_TEST(TestConvexOptimization, TestTrivialSDP) {
...
-}
-}  // namespace
-
 // Solve a semidefinite programming problem.
 // Find the common Lyapunov function for linear systems
 // xdot = Ai*x
 // The condition is
 // min 0
 // s.t P is positive semidefinite
 //     Ai'*P + P*Ai is positive semidefinite
```
Comment (RussTedrake) [inline drake/solvers/test/convex_optimization_test.cc]: oops.  should read negative semidefinite here.

**Pull request merged**